src/services/upower/core.cpp

```diff
@@ -0,0 +1,110 @@
+#include "core.hpp"
+
+#include <qcontainerfwd.h>
+#include <qdbusconnection.h>
+#include <qdbusconnectioninterface.h>
+#include <qdbusextratypes.h>
+#include <qdbuspendingcall.h>
+#include <qdbuspendingreply.h>
+#include <qdbusservicewatcher.h>
+#include <qlogging.h>
+#include <qloggingcategory.h>
+#include <qobject.h>
+#include <qqmllist.h>
+#include <qtmetamacros.h>
+
+#include "../../core/model.hpp"
+#include "dbus_app.h"
+#include "device.hpp"
+
+namespace qs::service::upower {
+
+Q_LOGGING_CATEGORY(logUPower, "quickshell.service.upower", QtWarningMsg);
+
+UPower::UPower(QObject* parent): QObject(parent) {
+	auto bus = QDBusConnection::systemBus();
+
+	if (!bus.isConnected()) {
+		qCWarning(logUPower) << "Could not connect to DBus. UPower service will not work.";
+		return;
+	}
+
+	this->app = new DBusUPowerApp("org.freedesktop.UPower", "/org/freedesktop/UPower", bus, this);
+
+	if (!this->app->isValid()) {
+		qCWarning(logUPower) << "Cannot connect to the UPower service.";
+		return;
+	}
+
+	this->registerExisting();
+}
+
+void UPower::registerExisting() {
+	qCDebug(logUPower) << "Registering existing devices.";
+
+	this->registerDevice("/org/freedesktop/UPower/devices/DisplayDevice");
+
+	auto pending = this->app->EnumerateDevices();
+	auto* call = new QDBusPendingCallWatcher(pending, this);
+
+	auto responseCallback = [this](QDBusPendingCallWatcher* call) {
+		const QDBusPendingReply<QList<QDBusObjectPath>> reply = *call;
+
+		if (reply.isError()) {
+			qCWarning(logUPower) << "Failed to enumerate devices:" << reply.error().message();
+		} else {
+			for (const QDBusObjectPath& devicePath: reply.value()) {
+				this->registerDevice(devicePath.path());
+			}
+		}
+
+		delete call;
+	};
+
+	QObject::connect(call, &QDBusPendingCallWatcher::finished, this, responseCallback);
+}
+
+void UPower::onDeviceReady() {
+	auto* device = qobject_cast<UPowerDevice*>(this->sender());
+
+	if (device->path() == "/org/freedesktop/UPower/devices/DisplayDevice") {
+		this->mDisplayDevice = device;
+		emit this->displayDeviceChanged();
+		qCDebug(logUPower) << "Display UPowerDevice" << device->path() << "ready";
+		return;
+	}
+
+	this->readyDevices.insertObject(device);
+	qCDebug(logUPower) << "UPowerDevice" << device->path() << "ready";
+}
+
+void UPower::onDeviceDestroyed(QObject* object) {
+	auto* device = static_cast<UPowerDevice*>(object); // NOLINT
+	this->readyDevices.removeObject(device);
+}
```

Comment:
Firstly add a debug log, secondly make sure mDisplayDevice is set to null if it was the one destroyed. Additionally remove it from `mDevices`.